### PR TITLE
rgw/notify: decouple add/remove_persistent_topic() from Manager

### PR DESCRIPTION
**Pull request by cbodley**
allow persistent topics to be added/removed even if no rgw::notify::Manager is running. hide the rados dependency behind new (hopefully temporary) sal::Driver interfaces

Fixes: https://tracker.ceph.com/issues/65668

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Review by kchheda3 — APPROVED** [review]
Thanks @cbodley

**Comment by cbodley** [comment]
oops
```
src/rgw/rgw_sal_dbstore.cc:1917:69: error: 'override' specifier is not allowed outside a class definition
                                    const std::string& topic_queue) override
                                                                    ^~~~~~~~
src/rgw/rgw_sal_dbstore.cc:1924:72: error: 'override' specifier is not allowed outside a class definition
                                       const std::string& topic_queue) override
                                                                       ^~~~~~~~
2 errors generated.
```

**Comment by yuvalif** [comment]
can you please add a test to `src/test/rgw/bucket_notification/test_bn.py:test_ps_s3_topic_admin_on_master` 
that calls: `admin(['topic', 'rm', '--topic',...)` on a persistent topic?

File: src/rgw/driver/rados/rgw_notify.cc [hunk 1]
```diff
@@ -117,7 +117,6 @@ void publish_commit_completion(rados_completion_t completion, void *arg) {
 };
 
 class Manager : public DoutPrefixProvider {
-  const size_t max_queue_size;
```
Comment (yuvalif) [inline src/rgw/driver/rados/rgw_notify.cc:120]: how is this change related?

Comment (cbodley) [reply src/rgw/driver/rados/rgw_notify.cc:120]: that member variable is unused after add_persistent_topic() was removed from the class

**Comment by cbodley** [comment]
> can you please add a test to `src/test/rgw/bucket_notification/test_bn.py:test_ps_s3_topic_admin_on_master` that calls: `admin(['topic', 'rm', '--topic',...)` on a persistent topic?

@yuvalif there are already 3 `topic rm` commands in the test, so i just added `persistent=true` to one of those existing topics. does that look ok?

**Review by yuvalif — APPROVED** [review]

**Comment by kchheda3** [comment]
> > can you please add a test to `src/test/rgw/bucket_notification/test_bn.py:test_ps_s3_topic_admin_on_master` that calls: `admin(['topic', 'rm', '--topic',...)` on a persistent topic?
> 
> @yuvalif there are already 3 `topic rm` commands in the test, so i just added `persistent=true` to one of those existing topics. does that look ok?

problem with these test cases is, admin delete topic will still return 0, as the error of not being able to delete the persistent queue is [swallowed](https://github.com/ceph/ceph/blob/main/src/rgw/rgw_pubsub.cc#L1145) and not reported
so actual test case would be to check if the queue is deleted or report error of failing to delete queue up ?

**Comment by yuvalif** [comment]
> > > can you please add a test to `src/test/rgw/bucket_notification/test_bn.py:test_ps_s3_topic_admin_on_master` that calls: `admin(['topic', 'rm', '--topic',...)` on a persistent topic?
> > 
> > 
> > @yuvalif there are already 3 `topic rm` commands in the test, so i just added `persistent=true` to one of those existing topics. does that look ok?
> 
> problem with these test cases is, admin delete topic will still return 0, as the error of not being able to delete the persistent queue is [swallowed](https://github.com/ceph/ceph/blob/main/src/rgw/rgw_pubsub.cc#L1145) and not reported so actual test case would be to check if the queue is deleted or report error of failing to delete queue up ?

I see. so, the problem with replying with an error is that the user will not be able to fix it. since the topic is already deleted, any retry of the deletion will see that the topic is not there anymore, and will not continue (and return ENOENT).
to fix that, we shoud probably reorder the topic deletion sequence, deleting the persistent queue before we delete the topic. in this case we should be able to return an error when the queue deletion fails.
lets do that in another PR.

**Comment by cbodley** [comment]
in testing, i've seen the same `FAIL: test pushing http s3 notification` failure twice in a row now. that branch includes the fixes from https://github.com/ceph/ceph/pull/56979. i opened https://tracker.ceph.com/issues/66033 with details

the deletes are processed by radosgw, not radosgw-admin, so i'm not sure how this pr could cause a regression there

**Comment by cbodley** [comment]
> the deletes are processed by radosgw, not radosgw-admin, so i'm not sure how this pr could cause a regression there

oh, i see the same failure on another branch too: https://qa-proxy.ceph.com/teuthology/cbodley-2024-05-15_12:34:15-rgw-wip-boost-asio-spawn-distro-default-smithi/7707122/teuthology.log

**Comment by kchheda3** [comment]
> > the deletes are processed by radosgw, not radosgw-admin, so i'm not sure how this pr could cause a regression there
> 
> oh, i see the same failure on another branch too: https://qa-proxy.ceph.com/teuthology/cbodley-2024-05-15_12:34:15-rgw-wip-boost-asio-spawn-distro-default-smithi/7707122/teuthology.log

the failure is not about topic delete, its about the delete event not getting delivered.
this PR is handling deletion of topic, and failure is about object delete event not getting delivered

**Comment by cbodley** [comment]
calling this a pass based on https://pulpito.ceph.com/cbodley-2024-05-14_15:39:36-rgw-wip-cbodley-testing-distro-default-smithi/, tracking the notification failure in https://tracker.ceph.com/issues/66033

**Comment by cbodley** [comment]
squid backport will conflict until https://github.com/ceph/ceph/pull/57470 merges
